crates/block-producer/src/mempool/mod.rs

```diff
@@ -274,55 +223,119 @@ impl Mempool {
     ///
-    /// Transactions are placed back in the queue.
+    /// Transactions are re-queued.
     #[instrument(target = COMPONENT, name = "mempool.rollback_batch", skip_all)]
     pub fn rollback_batch(&mut self, batch: BatchId) {
-        // Batch may already have been removed as part of a parent batches failure.
-        if !self.batches.contains(&batch) {
+        // Due to the distributed nature of the system, its possible that a proposed batch was
+        // already proven, or already reverted. This guards against this eventuality.
+        if !self.nodes.proposed_batches.contains_key(&batch) {
             return;
         }
 
-        let removed_batches =
-            self.batches.remove_batches([batch].into()).expect("Batch was not present");
+        // TODO(mirko): This will be somewhat complicated by the introduction of user batches
+        // since these don't have all inner txs present and therefore must at least partially
+        // revert their own tx descendents.
 
-        let transactions = removed_batches.values().flatten().copied().collect();
+        // Remove all descendents and reinsert their transactions. This is safe to do since
+        // a batch's state impact is the aggregation of its transactions.
+        let reverted = self.revert_subtree(NodeId::ProposedBatch(batch));
 
-        self.transactions
-            .requeue_transactions(transactions)
-            .expect("Transaction should requeue");
+        for (_, node) in reverted {
+            for tx in node.transactions() {
+                let tx = TransactionNode::new(Arc::clone(tx));
+                self.state.insert(NodeId::Transaction(tx.id()), &tx);
+                self.nodes.txs.insert(tx.id(), tx);
+            }
+        }
 
-        tracing::warn!(
-            %batch,
-            descendents=?removed_batches.keys(),
-            "Batch failed, dropping all inflight descendent batches, impacted transactions are back in queue."
-        );
         self.inject_telemetry();
     }
 
     /// Marks a batch as proven if it exists.
     #[instrument(target = COMPONENT, name = "mempool.commit_batch", skip_all)]
-    pub fn commit_batch(&mut self, batch: ProvenBatch) {
-        // Batch may have been removed as part of a parent batches failure.
-        if !self.batches.contains(&batch.id()) {
+    pub fn commit_batch(&mut self, proof: Arc<ProvenBatch>) {
+        // Due to the distributed nature of the system, its possible that a proposed batch was
+        // already proven, or already reverted. This guards against this eventuality.
+        let Some(proposed) = self.nodes.proposed_batches.remove(&proof.id()) else {
             return;
-        }
+        };
+
+        self.state.remove(&proposed);
+
+        let proven = proposed.to_proven_batch_node(proof);
+        self.state.insert(NodeId::ProvenBatch(proven.id()), &proven);
+        self.nodes.proven_batches.insert(proven.id(), proven);
```

Comment:
When a batch transitions from proposed to proven, we remove the proposed batch and reinsert the proven batch. This essentially re-associates all state it references (e.g. output notes it creates) from the `NodeId::ProposedBatch` (removed) to the new `NodeId::ProvenBatch` (inserted), right?

I'm asking because while they carry the same `BatchId` internally, the `Hash` implementation is different for these IDs, but if the above is correct, then that's fine. Maybe you can add a comment on `NodeId` that the Hash impl is only used for deduplication purposes or something like that, but also fine as is.

Reply:
> When a batch transitions from proposed to proven, we remove the proposed batch and reinsert the proven batch. This essentially re-associates all state it references (e.g. output notes it creates) from the NodeId::ProposedBatch (removed) to the new NodeId::ProvenBatch (inserted), right?

That's correct.

In practice, there is also a slight difference in the state of both because the proposed batch doesn't actually aggregate the txs state i.e. it doesn't erase notes, nor fold account updates together. But that's more an implementation detail, which could be changed if desired.